jni/lib/src/jni.dart

```diff
@@ -46,102 +41,58 @@ DynamicLibrary _loadJniHelpersLibrary(
-            " Did you call Jni.spawn?");
-      }
-      // If no error, save this singleton.
-      _instance = inst;
-    }
-    return _instance!;
-  }
+  /// Store dylibDir if any was used.
+  static String? _dylibDir;
 
-  /// Initialize instance from custom helper library path.
-  ///
-  /// On dart standalone, call this in new isolate before
-  /// doing getInstance().
+  /// Sets the directory where dynamic libraries are looked for.
+  /// On dart standalone, call this in new isolate before doing
+  /// any JNI operation.
   ///
   /// (The reason is that dylibs need to be loaded in every isolate.
   /// On flutter it's done by library. On dart standalone we don't
   /// know the library path.)
-  static void load({required String helperDir}) {
-    if (_instance != null) {
-      throw StateError('Fatal: a JNI instance already exists in this isolate');
-    }
-    final inst = Jni._(_loadJniHelpersLibrary(dir: helperDir), helperDir);
-    if (inst.getJavaVM() == nullptr) {
-      throw StateError("Fatal: No JVM associated with this process");
-    }
-    _instance = inst;
+  static void setDylibDir({required String dylibDir}) {
+    _dylibDir = dylibDir;
   }
 
-  /// Spawn an instance of JVM using JNI.
-  /// This instance will be returned by future calls to [getInstance]
+  /// Spawn an instance of JVM using JNI. This method should be called at the
+  /// beginning of the program with appropriate options, before other isolates
+  /// are spawned.
   ///
-  /// [helperDir] is path of the directory where the wrapper library is found.
+  /// [dylibDir] is path of the directory where the wrapper library is found.
   /// This parameter needs to be passed manually on __Dart standalone target__,
   /// since we have no reliable way to bundle it with the package.
   ///
   /// [jvmOptions], [ignoreUnrecognized], & [jniVersion] are passed to the JVM.
   /// Strings in [classPath], if any, are used to construct an additional
   /// JVM option of the form "-Djava.class.path={paths}".
-  static Jni spawn({
-    String? helperDir,
-    int logLevel = JniLogLevel.JNI_INFO,
+  static void spawn({
+    String? dylibDir,
     List<String> jvmOptions = const [],
     List<String> classPath = const [],
     bool ignoreUnrecognized = false,
     int jniVersion = JNI_VERSION_1_6,
   }) {
-    if (_instance != null) {
-      throw UnsupportedError("Currently only 1 VM is supported.");
+    _dylibDir = dylibDir;
+    final existVm = _bindings.GetJavaVM();
+    if (existVm != nullptr) {
+      throw JvmExistsException();
     }
-    final dylib = _loadJniHelpersLibrary(dir: helperDir);
-    final inst = Jni._(dylib, helperDir);
-    _instance = inst;
-    inst._bindings.SetJNILogging(logLevel);
     final jArgs = _createVMArgs(
       options: jvmOptions,
       classPath: classPath,
       version: jniVersion,
       ignoreUnrecognized: ignoreUnrecognized,
     );
-    inst._bindings.SpawnJvm(jArgs);
+    _bindings.SpawnJvm(jArgs);
     _freeVMArgs(jArgs);
```

Comment:
Use a single arena to allocate everything, with a using wrapping the whole spawn body. Then the whole `_freeVMArgs` isn't needed.